### PR TITLE
Fixed offset and limit query clause returning partial query results when used with distinct

### PR DESCRIPTION
**Pull request by kushagraThapar**
Currently, OFFSET and LIMIT clause continuation tokens are mixed with TOP clause continuation tokens. This breaks the scenario where query results need to be filtered further than the OFFSET and LIMIT values, like using DISTINCT on client side. This causes the Java SDK to return partial query results irrespective of the page size passed in (as it filters the first top results - because top is treated as same as limit)

To fix this, we need to separate the TOP and LIMIT clause continuation tokens similar to what we have in .Net. This PR separates them and fixes the issue. (Inspired by: https://msdata.visualstudio.com/CosmosDB/_git/CosmosDB?path=/Product/SDK/.net/Microsoft.Azure.Documents.Client/Query/ExecutionComponent/TakeDocumentQueryExecutionComponent.cs)

### CODE REVIEW

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DefaultDocumentQueryExecutionContext.java [hunk 1]
```diff
@@ -257,8 +257,12 @@ private static boolean isClientSideContinuationToken(String continuationToken) {
                 return true;
             }
 
-            ValueHolder<TakeContinuationToken> outTakeContinuationToken = new ValueHolder<>();
-            return TakeContinuationToken.tryParse(continuationToken, outTakeContinuationToken);
+            ValueHolder<TopContinuationToken> outTopContinuationToken = new ValueHolder<>();
+            if (TopContinuationToken.tryParse(continuationToken, outTopContinuationToken)) {
+                return true;
+            }
+            ValueHolder<LimitContinuationToken> outLimitContinuationToken = new ValueHolder<>();
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DefaultDocumentQueryExecutionContext.java:264]: I dont think we use this path anymore :-)! There are various other composite continuation tokens which could mark the continuation token as clientSideContinuationTokens.

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DistinctContinuationToken.java [hunk 2]
```diff
@@ -16,7 +16,7 @@ public class DistinctContinuationToken extends JsonSerializable {
     private static final String LAST_HASH_PROPERTY_NAME = "lastHash";
     private static final String SOURCE_TOKEN_PROPERTY_NAME = "sourceToken";
 
-    private static final Logger logger = LoggerFactory.getLogger(TakeContinuationToken.class);
+    private static final Logger logger = LoggerFactory.getLogger(DistinctContinuationToken.class);
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/DistinctContinuationToken.java:19]: Good catch 👍

File: sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/TakeDocumentQueryExecutionContext.java [hunk 3]
```diff
@@ -17,56 +18,92 @@
...
     }
 
-    public static <T> Flux<IDocumentQueryExecutionComponent<T>> createAsync(
-            BiFunction<String, PipelinedDocumentQueryParams<T>, Flux<IDocumentQueryExecutionComponent<T>>> createSourceComponentFunction,
-            int topCount,
-            int limit,
-            String topContinuationToken,
-            PipelinedDocumentQueryParams<T> documentQueryParams) {
-        TakeContinuationToken takeContinuationToken;
+    public static <T> Flux<IDocumentQueryExecutionComponent<T>> createTopAsync(BiFunction<String, PipelinedDocumentQueryParams<T>, Flux<IDocumentQueryExecutionComponent<T>>> createSourceComponentFunction,
```
Comment (mbhaskar) [inline sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/implementation/query/TakeDocumentQueryExecutionContext.java:35]: `createAsync` static method is a common pattern followed by all the QueryExecutionContext classes. Maybe you could have `createAsync` and switch between top or limit inside it?

**Review by mbhaskar — APPROVED** [review]
LGTM. Thanks for the changes. 
If possible, please add test for the limit + distinct scenario. QueryValidationTests would be a good place.

**Comment by kushagraThapar** [comment]
> LGTM. Thanks for the changes. If possible, please add test for the limit + distinct scenario. QueryValidationTests would be a good place.

@mbhaskar - added further tests and code improvements, thanks for your review, appreciate it.

**Comment by kkxuwork** [comment]
Hi @kushagraThapar When we use DISTINCT with ORDER BY and OFFSET 0 LIMT 100, the result is incorrect. Compared with without ORDER BY, there is a lot less data. How should we solve this problem? Thanks
For example, in the following two sql, the first one does not get 100 results, but the second one can get 100 results.
select u.name, u.age from user u where u.region = 'US' order by u.createTime offset 0 limit 100;
select u.name, u.age from user u where u.region = 'US' offset 0 limit 100;

**Comment by kushagraThapar** [comment]
thanks @kkxuwork for raising this issue. Can you please create a github issue with the information and how to replicate the issue with the schema if possible and doesn't break privacy information. We will work on it.

**Pull request merged**